core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java

```diff
@@ -677,57 +569,204 @@ public Decompressor getDecompressor() {
-           */
+          // The following statement is necessary to get around bugs in 0.18 where a decompressor is
+          // referenced after
+          // it's returned back to the codec pool.
           decompressor.reset();
         }
         return decompressor;
       }
-
       return null;
     }
 
-    public void returnDecompressor(Decompressor decompressor) {
+    /**
+     * Returns the given {@link Decompressor} to the codec cache if it is not null.
+     *
+     * @param decompressor
+     *          the decompressor to return
+     */
+    public void returnDecompressor(final Decompressor decompressor) {
       if (decompressor != null) {
-        log.debug("Returned a decompressor: {}", decompressor.hashCode());
+        LOG.debug("Returned a decompressor: {}", decompressor.hashCode());
         CodecPool.returnDecompressor(decompressor);
       }
     }
 
+    /**
+     * Returns the name of the compression algorithm.
+     *
+     * @return the name
+     */
     public String getName() {
-      return compressName;
+      return name;
     }
-  }
 
-  static Algorithm getCompressionAlgorithmByName(String compressName) {
-    Algorithm[] algos = Algorithm.class.getEnumConstants();
+    /**
+     * Initializes and returns a new codec with the specified buffer size if and only if the
+     * specified {@link AtomicBoolean} has a value of false, or returns the specified original coded
+     * otherwise.
+     *
+     * @param checked
+     *          whether or not the codec has already been checked
+     * @param bufferSize
+     *          the default buffer size for the new coded
+     * @param originalCodec
+     *          the original codec
+     * @return a codec
+     */
+    CompressionCodec initCodec(final AtomicBoolean checked, final int bufferSize,
+        final CompressionCodec originalCodec) {
+      if (!checked.get()) {
+        checked.set(true);
+        return createNewCodec(bufferSize);
+      }
+      return originalCodec;
+    }
 
-    for (Algorithm a : algos) {
-      if (a.getName().equals(compressName)) {
-        return a;
+    /**
+     * Returns a new {@link CompressionCodec} of the specified type, or the default type if no
+     * primary type is specified. If the specified buffer size is greater than 0, the specified
+     * buffer size configuration option will be updated in the codec's configuration with the buffer
+     * size.
+     *
+     * @param codecClazz
+     *          the codec type to initialize and return
+     * @param defaultClazz
+     *          the default codec type to return if codecClazz is null
+     * @param bufferSize
+     *          the buffer size to use (if greater than 0)
+     * @param bufferSizeConfigOpt
+     *          the configuration opt for the buffer size
+     * @return the initialized {@link CompressionCodec}, or null if the codec type was not found
+     */
+    CompressionCodec createNewCodec(final String codecClazz, final String defaultClazz,
```

Comment:
I'm a little bit confused about what `codecProp` would be communicating more clearly vs. `codecClazz`. Could you elaborate on this?

Personally, I think the `codecClazz` is clearer in terms of communicating that it's the type of codec that we want returned, especially in conjunction with the second argument `defaultClazz`, but I'd like to hear your thoughts.

Reply:
When I first started reviewing I thought the value of the variable would be a class name.  Looking around I finally figured out the value was a property name.  The most descriptive name would be `codeClazzProp`.  If you like `codecClazz` that is fine with me, naming is very subjective.   What is clear to me may not be clear to others.  Personally I found it confusing at first though.

Reply:
Ah, now I see, I made the mistake as well of thinking the value was a class name rather than a configuration property. I will change the name to `codecClazzProp`.